Add GitHub Actions workflow for publishing design system preview

aduth opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of DS-151

Closes DS-162

## Proposed Changes

* Adds GitHub Actions workflow to publish Automattic Design System reference site preview on pull request changes 

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Make it easier to preview changes, particularly for non-engineer stakeholders

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* The pull request should have a comment linking to a functioning preview site

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Add GitHub Actions workflow for publishing design system preview

aduth committed:
Limit Design System Docs workflow to edits on relevant paths

aduth committed:
Fix quoting on worker upload script

aduth committed:
Output URL for debugging

aduth commented:
This is working now, evidenced by the comment at https://github.com/Automattic/wp-calypso/pull/103778#issuecomment-2916646918 .

A few potential areas for improvement:

- Consolidating or eliminating comments with [the calypso.live preview comment](https://github.com/Automattic/wp-calypso/pull/103778#issuecomment-2916639036). Maybe we add a condition to disable the calypso.live comment if we know that we'll be publishing a design system reference site preview?
- Paths to watch: Is it sufficient to look for changes within the design system site content itself, or should we also generate a preview on any changes to `packages/components`? Any other files?

aduth committed:
Condition to avoid running actions on fork PRs

tyxla commented on apps/design-system-docs/wrangler.jsonc:
```diff
@@ -0,0 +1,8 @@
+{
+	"name": "preview",
+	"compatibility_date": "2025-05-28",
```
We might need to update some docs with regards to how/when this should be updated: 

https://developers.cloudflare.com/workers/configuration/compatibility-dates/#setting-compatibility-date

tyxla reviewed (commented):
This is looking great @aduth! 🚀 

> Consolidating or eliminating comments with https://github.com/Automattic/wp-calypso/pull/103778#issuecomment-2916639036. Maybe we add a condition to disable the calypso.live comment if we know that we'll be publishing a design system reference site preview?

SInce you talked about consolidating, is there any chance to combine this with the existing "Calypso Live" workflow, and then use the same comment instead of posting a new one? Right now, the separate comment is the only downside I can see.

> Paths to watch: Is it sufficient to look for changes within the design system site content itself, or should we also generate a preview on any changes to packages/components? Any other files?

As for paths to watch, would it be expensive if we build it for every PR, like the calypso.live workflow does? If yes, we might want to optionally add that part of the comment only when there are changes inside the DS app (should be sufficient for now IMO).

aduth commented:
> > Consolidating or eliminating comments with [#103778 (comment)](https://github.com/Automattic/wp-calypso/pull/103778#issuecomment-2916639036). Maybe we add a condition to disable the calypso.live comment if we know that we'll be publishing a design system reference site preview?
> 
> SInce you talked about consolidating, is there any chance to combine this with the existing "Calypso Live" workflow, and then use the same comment instead of posting a new one? Right now, the separate comment is the only downside I can see.

Aside from the technical complexity of trying to manage a single comment with two separate Actions workflows, I wondered if there'd really be a need for both. For example, if I'm working on a pull request that updates content of the reference site, do I even need a link to the calypso.live preview? And vice-versa. I think the latter part of the comment is interesting though since some code changes can potentially have indirect effects on the reference site (e.g. components updates).

Worst-case, since the calypso.live comment content is pretty straight-forward, we could duplicate that into this action, so it appears as if it's the same comment. And then ensure the default calypso.live doesn't comment in cases where we'd expect this one to be.

> > Paths to watch: Is it sufficient to look for changes within the design system site content itself, or should we also generate a preview on any changes to packages/components? Any other files?
> 
> As for paths to watch, would it be expensive if we build it for every PR, like the calypso.live workflow does? If yes, we might want to optionally add that part of the comment only when there are changes inside the DS app (should be sufficient for now IMO).

To me it's less about being expensive, and more about whether it's relevant or creating additional unnecessary noise in a pull request. Maybe less of an issue if we consolidate the comments.

aduth committed:
Avoid duplicate comment for calypso.live preview

aduth commented:
The latest round of changes has created a new rabbit-hole where I'm now discovering we have two separate CI workflows for managing Calypso.live comments, both [in TeamCity](https://github.com/Automattic/wp-calypso/blob/b2887aa0aa9b93cbd0da67358494fedecc45c516/.teamcity/_self/projects/WebApp.kt#L203-L255) and [in GitHub Actions](https://github.com/Automattic/wp-calypso/blob/trunk/.github/workflows/calypso-live.yml). This is explained in #74501:

>Why do we have two? Well, there is an edge case where the TeamCity job will execute before you have a PR created (because it triggers on commit push.) In this scenario, you’d never get a comment about calypso.live. Additionally, it can be helpful for GitHub actions to add the comment immediately, when the PR is opened, rather than waiting several minutes for the TeamCity job to finish.
>
>But this only impacts PR creation. After that, the TeamCity job will maintain a single comment, keeping it updated. This job also deletes any existing comment from GitHub, to avoid having duplicates.

This feels a little strange to me, where we could be using _just_ GitHub Actions, especially given how simple/static the calypso.live content is. The only issue is that the current action is configured to create, but not update the comment. Since the action being introduced here is capable of doing both using [the `create-or-update-comment` action](https://github.com/peter-evans/create-or-update-comment), it might be worth consolidating to use a common approach.

But that also inflates the scope of this pull request, so I'm somewhat inclined to revert the latest change, merge this as-is, and then address the consolidation of live commenting behavior in a follow-up pull request. It might also be a good opportunity to revisit the formatting of the calypso.live comment, which is more complex to support QR codes, but those QR codes are broken anyways.

aduth committed:
Revert "Avoid duplicate comment for calypso.live preview"

This reverts commit d74f7fb4a34603fac7beeed7fce7ad1d2dc9aaf1.

tyxla commented:
> I'm somewhat inclined to revert the latest change, merge this as-is, and then address the consolidation of live commenting behavior in a follow-up pull request. It might also be a good opportunity to revisit the formatting of the calypso.live comment, which is more complex to support QR codes, but those QR codes are broken anyways.

As long as we don't cause extra noise by publishing comments in unrelated PRs, I wouldn't be opposed to that 👍

aduth commented:
> > I'm somewhat inclined to revert the latest change, merge this as-is, and then address the consolidation of live commenting behavior in a follow-up pull request. It might also be a good opportunity to revisit the formatting of the calypso.live comment, which is more complex to support QR codes, but those QR codes are broken anyways.
> 
> As long as we don't cause extra noise by publishing comments in unrelated PRs, I wouldn't be opposed to that 👍

Correct, it's still limited to pull requests modifying either the workflow itself, or files in `app/design-system-docs`.

aduth commented:
Ah, yeah, I had switched to using `github.event.pull_request.head.sha` instead of `github.sha` in the now-reverted 8348f31fe7be96192595bead5b43b7dba04880b8, since it's the latest commit in the branch rather than a generated merge commit of what would be merged to `trunk`. I think that'd address what you're seeing. I'll bring that specific change back before merging.

aduth committed:
Use pull request HEAD SHA for commit reference

See: https://github.com/Automattic/wp-calypso/pull/103778#issuecomment-2921967703

aduth commented:
> > I'm somewhat inclined to revert the latest change, merge this as-is, and then address the consolidation of live commenting behavior in a follow-up pull request. It might also be a good opportunity to revisit the formatting of the calypso.live comment, which is more complex to support QR codes, but those QR codes are broken anyways.
> 
> As long as we don't cause extra noise by publishing comments in unrelated PRs, I wouldn't be opposed to that 👍

Follow-up (draft) for consolidated pull request comments: #103848
